astro/src/content/blog/backend-for-frontend-security-architecture.mdx

```diff
@@ -0,0 +1,511 @@
+
+The only solution is to never allow tokens to enter the browser context. That's where Backend-for-Frontend comes in.
+
+## Understanding Backend-for-Frontend Architecture
+
+Instead of treating the browser as a secure environment for token storage, BFF acknowledges the reality that the browser is a hostile environment that tokens should never be exposed to.
+
+In a BFF architecture, your backend becomes a secure vault that holds all OAuth tokens - access tokens, refresh tokens, and ID tokens. The browser only receives an opaque session identifier in an httpOnly cookie, which JavaScript cannot read or access. This single architectural decision defeats entire categories of attacks.
+
+The BFF has three core responsibilities:
+
+1. Acting as a confidential OAuth client - The BFF authenticates with the authorization server using a client secret, something a browser can never securely do.
+2. Managing OAuth tokens in server-side sessions - All tokens stay on your backend, never exposed to potentially compromised JavaScript.
+3. Proxying API requests - The browser makes requests to your BFF, which adds the appropriate access token before forwarding to resource servers.
+
+### Architecture Comparison
+
+Let's visualize the difference between traditional SPA authentication and BFF:
+
+{/* TODO: Image comparing traditional SPA authentication and BFF architecture */}
+
+In BFF, there are simply no tokens for malicious code to steal.
+
+### How BFF Prevents Attacks
+
+| Attack Type                  | Traditional SPA (Tokens in Browser)      | BFF (Tokens on Backend)              |
+| ---------------------------- | ---------------------------------------- | ------------------------------------ |
+| Single Execution Token Theft | ✅ Steals tokens from localStorage       | ❌ No tokens in browser to steal     |
+| Persistent Token Theft       | ✅ Repeatedly steals tokens              | ❌ No tokens in browser to steal     |
+| Acquisition of New Tokens    | ✅ Uses hidden iframes to get new tokens | ❌ No browser tokens, no silent auth |
+
+The only remaining attack vector is request proxying (5.1.4) - the attacker can make requests to your BFF while the user's browser is open. But they can't exfiltrate tokens for offline use, can't persist access beyond the session, and can't escalate privileges.
+
+### Security Wins
+
+Here are the concrete security benefits of BFF:
+
+#### 1. Confidential Client Authentication
+
+```javascript
+// Backend can securely use client credentials
+const tokenResponse = await fetch(tokenEndpoint, {
+  method: 'POST',
+  body: new URLSearchParams({
+    grant_type: 'authorization_code',
+    code: authCode,
+    client_id: process.env.CLIENT_ID,
+    client_secret: process.env.CLIENT_SECRET, // ← Browser can't do this
```

Comment:
Given that there has been some kerfuffle about ENV variables not being safe from AI, let's just use `clientID` and `clientSecret` as variables to not claim an "opinion" about this in this particular article.